File: measure-android/measure/src/main/java/sh/measure/android/config/ConfigLoader.kt
package sh.measure.android.config

internal interface ConfigLoader {
    /**
     * Returns the cached config synchronously, if available. Returns `null` if cached config is
     * unavailable or failed to load.
     */
    fun getCachedConfig(): Config?

    /**
     * Fetches a fresh config from the server asynchronously and calls [onSuccess] with the result,
     * if successful. Ignores the result if the fetch fails.
     */
    fun getNetworkConfig(onSuccess: (Config) -> Unit)
}

internal class ConfigLoaderImpl : ConfigLoader {
    override fun getCachedConfig(): Config? {
        // TODO:  Load the cached config from disk.
        return null
    }

    override fun getNetworkConfig(onSuccess: (Config) -> Unit) {
        // TODO: fetch the config from the server, write it to disk and call onSuccess.
    }
}

File: ios/MeasureSDK/MeasureSDK/Config/ConfigLoader.swift
//
//  ConfigLoader.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 26/08/24.
//

import Foundation

/// A protocol that defines the methods for loading configuration data either a cache or a network source.
protocol ConfigLoader {
    func getCachedConfig() -> Config?
    func getNetworkConfig(onSuccess: @escaping (Config) -> Void)
}

/// A base implementation of the `ConfigLoader` protocol.
struct BaseConfigLoader: ConfigLoader {
    /// Returns the cached configuration if available.
    /// - Returns: Optional cached `Confg`
    func getCachedConfig() -> Config? {
        // swiftlint:disable todo
        // TODO: Load the cached config from disk.
        return nil
    }

    /// Fetches a fresh configuration from the server.
    /// - Parameter onSuccess: A closure that returns a server `Config`
    func getNetworkConfig(onSuccess: @escaping (Config) -> Void) {
        // TODO: Fetch the config from the server, write it to disk, and call onSuccess.
        // swiftlint:enable todo
    }
}
